# One Vs Rest: Logistic Regression

#### On their own, logistic regressions are only binary classifiers, meaning they cannot handle target vectors with more than two classes. However, there are clever extensions to logistic regression to do just that. In one-vs-rest logistic regression (OVR) a separate model is trained for each class predicted whether an observation is that class or not (thus making it a binary classification problem). It assumes that each classification problem (e.g. class 0 or not) is independent. Consider that output which gives the highest probabilty.

In [2]:
# Load libraries
from sklearn.linear_model import LogisticRegression
from sklearn import datasets
from sklearn.preprocessing import StandardScaler

In [3]:
# Load data
iris = datasets.load_iris()
X = iris.data
y = iris.target

In [4]:
# Standarize features
scaler = StandardScaler()
X_std = scaler.fit_transform(X)

In [5]:
# Create one-vs-rest logistic regression object
clf = LogisticRegression(random_state=0, multi_class='ovr')

In [6]:
# Train model
model = clf.fit(X_std, y)

C:\Users\sjhanwar\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [7]:
# Create new observation
new_observation = [[.5, .5, .5, .5]] #  its just like x-test

In [8]:
# Predict class
model.predict(new_observation)

array([2])

In [9]:
# View predicted probabilities
model.predict_proba(new_observation)

array([[0.08401839, 0.29392375, 0.62205786]])

## We go ahead with output having probabilty of 0.622